compiler/rustc_ty_utils/src/opaque_types.rs

```diff
@@ -0,0 +1,251 @@
+use rustc_data_structures::fx::FxHashSet;
+use rustc_errors::ErrorGuaranteed;
+use rustc_hir::{def::DefKind, def_id::LocalDefId};
+use rustc_middle::ty::util::{CheckRegions, NotUniqueParam};
+use rustc_middle::ty::{self, Ty, TyCtxt, TypeVisitableExt};
+use rustc_middle::ty::{TypeSuperVisitable, TypeVisitable, TypeVisitor};
+use rustc_span::{Span, DUMMY_SP};
+use rustc_type_ir::AliasKind;
+use std::ops::ControlFlow;
+
+use crate::errors::{DuplicateArg, LateBoundInOpaque, NotParam};
+
+struct OpaqueTypeCollector<'tcx> {
+    tcx: TyCtxt<'tcx>,
+    opaques: Vec<ty::Binder<'tcx, Ty<'tcx>>>,
+    /// The `DefId` of the impl block if we are looking for opaque types
+    /// defined by an associated function.
+    parent: Option<LocalDefId>,
+    /// Where we are collecting the opaque types.
+    span: Span,
```

Comment:
I feel like it may be clearer to not precompute these two and just pass down the def id of the thing we're looking in...